java/arcs/sdk/wasm/WasmParticle.kt

```diff
@@ -0,0 +1,188 @@
+     * synchronization.
+     *
+     * @param name Name of the handle
+     * @param canRead Mark handle with read access
+     * @param canWrite Mark handle with write access
+     * @return The name-associated handle, or null
+     * @see [registerHandle]
+     * @see [onHandleSync]
+     */
+    @Suppress("UNUSED_PARAMETER")
+    fun connectHandle(name: String, canRead: Boolean, canWrite: Boolean): Handle? {
+        handles[name]?.let {
+            if (canRead) {
+                toSync.add(it)
+            }
+            return it
+        }
+
+        log("Handle $name not registered")
+        return null
+    }
+
+    /**
+     * Register a reaction to an event.
+     *
+     * Particle templates may emit events, usually from user actions.
+     *
+     * @param name The name of the triggered event
+     * @param handler A callback (consumer) in reaction to the event
+     */
+    fun eventHandler(name: String, handler: (Map<String, String>) -> Unit) {
+        eventHandlers[name] = handler
+    }
+
+    /**
+     * Trigger an event.
+     *
+     * Will target registered events, if present. Will always initiate rendering when called.
+     *
+     * @param slotName Slot that the event is associated with; likely `root`
+     * @param eventName Name of the event to trigger
+     * @param eventData Data associated with the event; will be passed into event handler
+     * @see [eventHandler]
+     */
+    open fun fireEvent(slotName: String, eventName: String, eventData: Map<String, String>) {
+        eventHandlers[eventName]?.invoke(eventData)
+        renderOutput()
+    }
+
+    /** @param handle Handle to synchronize */
+    fun sync(handle: Handle) {
+        toSync.remove(handle)
+        onHandleSync(handle, toSync.isEmpty())
+    }
+
+    /**
+     * React to handle updates.
+     *
+     * Called for handles when change events are received from the backing store. Default action is to trigger
+     * rendering.
+     *
+     * @param handle Singleton or Collection handle
+     */
+    open fun onHandleUpdate(handle: Handle) = Unit
+
+    /**
+     * React to handle synchronization.
+     *
+     * Called for handles that are marked for synchronization at connection, when they are updated with the full model
+     * of their data. This will occur once after setHandles() and any time thereafter if the handle is resynchronized.
+     * Default action is to trigger rendering.
+     *
+     * @param handle Singleton or Collection handle
+     * @param allSynced flag indicating if all handles are synchronized
+     */
+    open fun onHandleSync(handle: Handle, allSynced: Boolean) = Unit
+
+    /** Rendering through UiBroker */
+    fun renderOutput() {
+        val slotName = ""
+        val template = getTemplate(slotName)
+        val model = populateModel(slotName)?.let {
+            StringEncoder().encodeDictionary(it).toNullTermByteArray()
+        }
+        WasmRuntimeClient.onRenderOutput(this, template, model)
```

Comment:
Should this call be ignored of the model is null?

Reply:
I'm not sure, I think it's ok for the model to be null